In [1]:
# Nhập thư viện
# thư viện để phân tích dữ liệu
import pandas as pd
# thư viện để làm việc với chuỗi và số (numpy: numerical python)
import numpy as np

In [3]:
# đọc dữ liệu

df = pd.read_csv('OnlineRetail.csv', encoding = "ISO-8859-1")

# xem 5 dòng đầu dữ liệu
df.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

### Tìm hiểu cấu trúc bộ dữ liệu: Số lượng quan sát, số thuộc tính trong bộ dữ liệu, thang đo tương ứng của các thuộc tính


In [4]:
# số lượng dòng, cột của dữ liệu

df.shape

(541909, 8)

Dữ liệu có 541909 dòng, 8 cột

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


Bộ dữ liệu chứa 541909 dòng dữ liệu và 8 cột, chỉ có cột CustomerID không có đầy đủ thông tin (406829 non-null <541909 )

InvoiceNo, StockCode, Description, InvoiceDate, country: là thuộc tính định tính, có thang đo định danh

CustomerID mặc dù có kiểu số (float64), nhưng các con số này chỉ có ý nghĩa phân biệt khác hàng này với khách hàng khác –> thuộc tính định danh


Quantity, UnitPrice: thuộc tính định lượng, thang đo ratio 

### Hãy giải đáp các thắc mắc sau:Công ty bán hàng do bao nhiêu nước sản xuất

In [15]:
df['Country'].unique()

array(['United Kingdom', 'France', 'Australia', 'Netherlands', 'Germany',
       'Norway', 'EIRE', 'Switzerland', 'Spain', 'Poland', 'Portugal',
       'Italy', 'Belgium', 'Lithuania', 'Japan', 'Iceland',
       'Channel Islands', 'Denmark', 'Cyprus', 'Sweden', 'Austria',
       'Israel', 'Finland', 'Bahrain', 'Greece', 'Hong Kong', 'Singapore',
       'Lebanon', 'United Arab Emirates', 'Saudi Arabia',
       'Czech Republic', 'Canada', 'Unspecified', 'Brazil', 'USA',
       'European Community', 'Malta', 'RSA'], dtype=object)

In [16]:
df['Country'].unique().size

38

In [13]:
country = df['Country'].unique()
print ("số lượng các quốc gia: " + str(country.size))

số lượng các quốc gia: 38


In [12]:
df['Country'].value_counts()

Country
United Kingdom          495478
Germany                   9495
France                    8557
EIRE                      8196
Spain                     2533
Netherlands               2371
Belgium                   2069
Switzerland               2002
Portugal                  1519
Australia                 1259
Norway                    1086
Italy                      803
Channel Islands            758
Finland                    695
Cyprus                     622
Sweden                     462
Unspecified                446
Austria                    401
Denmark                    389
Japan                      358
Poland                     341
Israel                     297
USA                        291
Hong Kong                  288
Singapore                  229
Iceland                    182
Canada                     151
Greece                     146
Malta                      127
United Arab Emirates        68
European Community          61
RSA                         58


### Tổng số lượng đơn hàng bán ra, tổng doanh thu


In [19]:
# Tính tổng số lượng đơn hàng bán ra
df['InvoiceNo'].unique().size

25900

In [27]:
# tính tổng doanh thu

# tạo cột doanh thu
df['Revenu'] = df['Quantity'] * df['UnitPrice']
# tính tổng doanh thu dựa trên cột doanh thu vừa tạo
print(f'Tổng doanh thu:') 
df['Revenu'].sum()


Tổng doanh thu:


9747747.933999998

### Top 10 mặt hàng có số lượng bán ra lớn nhất

 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object -> thiếu dữ liệu
 3   Quantity     541909 non-null  int64  

In [33]:
# Nhóm số lượng theo StockCode: mã hàng

# Quantity: Số lương
# StockCode: mã hàng
df1 = df.groupby('StockCode')['Quantity'].sum()
df1


StockCode
10002           1037
10080            495
10120            193
10123C           -13
10123G           -38
                ... 
gift_0001_20      20
gift_0001_30      37
gift_0001_40       3
gift_0001_50       4
m                  1
Name: Quantity, Length: 4070, dtype: int64

In [36]:
# sau đó tìm ra 10 mặt hàng có số lượng bán ra lớn nhất 

df2 = df1.sort_values(ascending=False)
df2.head(10)

StockCode
22197     56450
84077     53847
85099B    47363
85123A    38830
84879     36221
21212     36039
23084     30646
22492     26437
22616     26315
21977     24753
Name: Quantity, dtype: int64

### Top 10 mặt hàng có doanh thu lớn nhất

In [37]:
df.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  Revenu  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom   15.30  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom   20.34  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom   22.00  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom   20.34  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom   20.34

In [38]:
# Nhóm doanh thu theo StockCode: mã hàng

# Quantity: Số lương
# StockCode: mã hàng
df3 = df.groupby('StockCode')['Revenu'].sum()
df3


StockCode
10002           759.89
10080           119.09
10120            40.53
10123C            3.25
10123G            0.00
                 ...  
gift_0001_20    167.05
gift_0001_30    175.53
gift_0001_40    100.70
gift_0001_50    167.56
m                 2.55
Name: Revenu, Length: 4070, dtype: float64

In [39]:
# sau đó tìm ra 10 mặt hàng có số doanh thu lớn nhất 

df4 = df3.sort_values(ascending=False)
df4.head(10)

StockCode
DOT       206245.48
22423     164762.19
47566      98302.98
85123A     97894.50
85099B     92356.03
23084      66756.59
POST       66230.64
22086      63791.94
84879      58959.73
79321      53768.06
Name: Revenu, dtype: float64

In [40]:
quantity_product = df.groupby(['StockCode', 'Description'])['Revenu'].sum().sort_values(ascending= False)
quantity_product.head(10)

StockCode  Description                       
DOT        DOTCOM POSTAGE                        206245.48
22423      REGENCY CAKESTAND 3 TIER              164762.19
47566      PARTY BUNTING                          98302.98
85123A     WHITE HANGING HEART T-LIGHT HOLDER     97715.99
85099B     JUMBO BAG RED RETROSPOT                92356.03
23084      RABBIT NIGHT LIGHT                     66756.59
POST       POSTAGE                                66230.64
22086      PAPER CHAIN KIT 50'S CHRISTMAS         63791.94
84879      ASSORTED COLOUR BIRD ORNAMENT          58959.73
79321      CHILLI LIGHTS                          53768.06
Name: Revenu, dtype: float64

Chúng ta nhìn thấy rằng, mặt hàng bán ra với số lượng nhiều nhất, không phải mặt hàng mang lại doanh thu lớn nhất 

Có một số mặt hàng như 85099B – JUMBO BAG RED RETROSPOT và 23084 RABBIT NIGHT LIGHT nằm trong top 10 cả về số lượng lẫn doanh thu